CUDADataFormats/SiPixelDigi/interface/SiPixelDigisCUDA.h

```diff
@@ -26,57 +27,16 @@ class SiPixelDigisCUDA {
   uint32_t nModules() const { return nModules_h; }
   uint32_t nDigis() const { return nDigis_h; }
 
-  uint16_t *xx() { return xx_d.get(); }
-  uint16_t *yy() { return yy_d.get(); }
-  uint16_t *adc() { return adc_d.get(); }
-  uint16_t *moduleInd() { return moduleInd_d.get(); }
-  int32_t *clus() { return clus_d.get(); }
-  uint32_t *pdigi() { return pdigi_d.get(); }
-  uint32_t *rawIdArr() { return rawIdArr_d.get(); }
+  cms::cuda::host::unique_ptr<uint16_t[]> copyAllToHostAsync(cudaStream_t stream) const;
 
-  uint16_t const *xx() const { return xx_d.get(); }
-  uint16_t const *yy() const { return yy_d.get(); }
-  uint16_t const *adc() const { return adc_d.get(); }
-  uint16_t const *moduleInd() const { return moduleInd_d.get(); }
-  int32_t const *clus() const { return clus_d.get(); }
-  uint32_t const *pdigi() const { return pdigi_d.get(); }
-  uint32_t const *rawIdArr() const { return rawIdArr_d.get(); }
-
-  cms::cuda::host::unique_ptr<uint16_t[]> adcToHostAsync(cudaStream_t stream) const;
-  cms::cuda::host::unique_ptr<int32_t[]> clusToHostAsync(cudaStream_t stream) const;
-  cms::cuda::host::unique_ptr<uint32_t[]> pdigiToHostAsync(cudaStream_t stream) const;
-  cms::cuda::host::unique_ptr<uint32_t[]> rawIdArrToHostAsync(cudaStream_t stream) const;
-
-  class DeviceConstView {
-  public:
-    __device__ __forceinline__ uint16_t xx(int i) const { return __ldg(xx_ + i); }
-    __device__ __forceinline__ uint16_t yy(int i) const { return __ldg(yy_ + i); }
-    __device__ __forceinline__ uint16_t adc(int i) const { return __ldg(adc_ + i); }
-    __device__ __forceinline__ uint16_t moduleInd(int i) const { return __ldg(moduleInd_ + i); }
-    __device__ __forceinline__ int32_t clus(int i) const { return __ldg(clus_ + i); }
-
-    uint16_t const *xx_;
-    uint16_t const *yy_;
-    uint16_t const *adc_;
-    uint16_t const *moduleInd_;
-    int32_t const *clus_;
-  };
-
-  const DeviceConstView *view() const { return view_d.get(); }
+  SiPixelDigisCUDASOAView *view() { return m_view.get(); }
+  SiPixelDigisCUDASOAView const *view() const { return m_view.get(); }
 
 private:
   // These are consumed by downstream device code
-  cms::cuda::device::unique_ptr<uint16_t[]> xx_d;         // local coordinates of each pixel
-  cms::cuda::device::unique_ptr<uint16_t[]> yy_d;         //
-  cms::cuda::device::unique_ptr<uint16_t[]> adc_d;        // ADC of each pixel
-  cms::cuda::device::unique_ptr<uint16_t[]> moduleInd_d;  // module id of each pixel
-  cms::cuda::device::unique_ptr<int32_t[]> clus_d;        // cluster id of each pixel
-  cms::cuda::device::unique_ptr<DeviceConstView> view_d;  // "me" pointer
+  cms::cuda::device::unique_ptr<uint16_t[]> m_store;
```

Comment:
Just to be on the safe side I would allocate uint32_t (and then use a uint16_t pointer anyhow).
In reality as long as we use the CashedAllocator is absolutely irrelevant (alignment guaranteed).
so OK keep it as such (just reverse the components as suggested below)